commit bd5503e6e5cc4df90713934aa621706007f99e18
Author: Eric Anholt <eric@anholt.net>
Date:   Mon Apr 17 09:26:03 2017 -0700

    drm/vc4: Fix refcounting of runtime PM get if it errors out.
    
    We were returning without decrementing if the error happened, meaning
    that at the next submit we wouldn't try to bring up the power domain.
    
    Signed-off-by: Eric Anholt <eric@anholt.net>
    Link: http://patchwork.freedesktop.org/patch/msgid/20170417162603.12726-1-eric@anholt.net
    Reviewed-by: Sean Paul <seanpaul@chromium.org>
    (cherry picked from commit 925d05e1f825db9490da33afe35bd5383d301e97)

diff --git a/drivers/gpu/drm/vc4/vc4_gem.c b/drivers/gpu/drm/vc4/vc4_gem.c
index 32e17b384fbc..994295fe1084 100644
--- a/drivers/gpu/drm/vc4/vc4_gem.c
+++ b/drivers/gpu/drm/vc4/vc4_gem.c
@@ -876,13 +876,16 @@ vc4_submit_cl_ioctl(struct drm_device *dev, void *data,
 	}
 
 	mutex_lock(&vc4->power_lock);
-	if (vc4->power_refcount++ == 0)
+	if (vc4->power_refcount++ == 0) {
 		ret = pm_runtime_get_sync(&vc4->v3d->pdev->dev);
-	mutex_unlock(&vc4->power_lock);
-	if (ret < 0) {
-		kfree(exec);
-		return ret;
+		if (ret < 0) {
+			mutex_unlock(&vc4->power_lock);
+			vc4->power_refcount--;
+			kfree(exec);
+			return ret;
+		}
 	}
+	mutex_unlock(&vc4->power_lock);
 
 	exec->args = args;
 	INIT_LIST_HEAD(&exec->unref_list);
